apis/apiextensions/v1beta1/composition_types.go

```diff
@@ -133,35 +212,145 @@ type ReadinessCheck struct {
-// after piping the value that is at FromFieldPath of the target resource through
-// transformers.
+// patchSetMap stores a list of patches from a set, against the Set ID.
+type patchSetMap map[string]Patch
+
+// PatchType defines the usable inbuilt patch types
+type PatchType string
+
+// Predefined PatchTypes.
+const (
+	PatchTypeFromCompositeFieldPath PatchType = "FromCompositeFieldPath" // Default
+	PatchTypePatchSet               PatchType = "PatchSet"
+)
+
+// Patch objects are applied between the composite and composed resources.
+// Their exact behaviour depends on the Type selected.
+// If the given type does not match one of the standard patch types, then
+// it may refer to a customPatchSet defined on the Composition. In this case,
+// the referenced patch will be dereferenced into the patch list in its' place.
 type Patch struct {
+	// Type can be a standard PatchSet value, or a 'Name' which refers
+	// to a custom patch type defined on the composition.
+	// +optional
+	// +kubebuilder:validation:Enum=FromCompositeFieldPath;PatchSet
+	// +kubebuilder:default=FromCompositeFieldPath
+	Type PatchType `json:"type,omitempty"`
+
+	// PatchSet is a patch type that allows multiple nested patches
+	// +optional
+	*PatchSet `json:",inline"`
+
+	// PatchFromCompositeFieldPath is the 'standard' patch type that maps a field
+	// from the composite resource to the composed resource, with optional
+	// transformations.
+	// +optional
+	*PatchFromCompositeFieldPath `json:",inline"`
+}
 
+// Apply applies a patch between the from and to objects. Depending on the
+// patch type, this may be from the composite resource to the composed resource,
+// or vice-versa.
+func (c *Patch) Apply(from, to runtime.Object) error {
+	var patcher interface {
+		Apply(from, to runtime.Object) error
+	}
+	switch c.Type {
+	case PatchTypeFromCompositeFieldPath:
+		patcher = c.PatchFromCompositeFieldPath
+	case PatchTypePatchSet:
+		patcher = c.PatchSet
+	default:
+		return errors.Errorf(errInvalidPatchType, c.Type)
+	}
+
+	// An interface equals nil only if both the type and value are nil. Above,
+	// even if t.<Type> is nil, its type is assigned to "transformer" but we're
+	// interested in whether only the value is nil or not.
+	if reflect.ValueOf(patcher).IsNil() {
+		return errors.Errorf(errPatchConfigMissing, c.Type)
+	}
+	return patcher.Apply(from, to)
+}
+
+// PatchSetCompatiblePatches embeds the valid PatchTypes which can be nested within
+// a PatchSet.
+type PatchSetCompatiblePatches struct {
+	*PatchFromCompositeFieldPath `json:",inline"`
+}
+
+// PatchSet defines a list of patches as a single grouping.
+// It can contain all other patch types except itself.
+type PatchSet struct {
+	// ID is the unique identifier of the PatchSet. ID is required when
+	// defining a new PatchSet, and also refers to a PatchSet from within
+	// the list of Patches on a resource.
+	// +optional
+	ID *string `json:"id,omitempty"`
```

Comment:
Is there any case where `id` is optional? Could we require it?

Reply:
The `PatchSet` type is used to both define a patch set in the `CompositionSpec` as well as reference it from the `Patch`. AIUI it needs to be optional when defined in the `Patch` otherwise it'll need to be set even when the `Type` is not `PatchSet`.

The `ID` field is currently checked at line 126 when the patch sets are mapped, but I suppose we could have another type def (`PatchSetDefinition` 🤷‍♂️) that requires the ID and is then converted to a normal `PatchSet` when mapped?

Reply:
Got it - I think leaving it as is makes sense.